[lldb] Allow languages to filter breakpoints set by line

felipepiovezan opened the pull request:
Some languages may create artificial functions that have no real user code, even though there is line table information for them. One such case is with coroutine code that receives the CoroSplitter transformation in LLVM IR. That code transformation creates many different Functions, cloning one Instruction into many Instructions in many different Functions and copying the associated debug locations.

It would be difficult to make that pass delete debug locations of cloned instructions in a language agnostic way (is it even possible?), but LLDB can ignore certain locations by querying its Language APIs and having it decide based on, for example, mangling information.

felipepiovezan committed:
[lldb] Allow languages to filter breakpoints set by line

Some languages may create artificial functions that have no real user code, even
though there is line table information for them. One such case is with coroutine
code that receives the CoroSplitter transformation in LLVM IR. That code
transformation creates many different Functions, cloning one Instruction into
many Instructions in many different Functions and copying the associated debug
locations.

It would be difficult to make that pass delete debug locations of cloned
instructions in a language agnostic way (is it even possible?), but LLDB can
ignore certain locations by querying its Language APIs and having it decide
based on, for example, mangling information.

jimingham commented:
LGTM except you should have a little bit of doc describing what "interesting" means.  On the face of it, this seems a little silly.  Why would the compiler emit "uninteresting" line table entries?  Your example motivation the PR makes that clear, that or something like in the docs would help motivate this.


Also, are these 100% uninteresting line table entries?  At present, there's no way to get lldb to not filter them out, so they either really need to be totally uninteresting, or we need a setting to turn them back on again.

felipepiovezan commented:
> Also, are these 100% uninteresting line table entries? At present, there's no way to get lldb to not filter them out, so they either really need to be totally uninteresting, or we need a setting to turn them back on again.

They should be 100% uninteresting, but I agree that it would be nice to have a mechanism allowing these to be re-enabled.
Do you think it is better to have a setting for this, or could we perhaps add the breakpoint but keep it disabled by default (is there precedent for this?) ?

jimingham commented:
We don't do the latter anywhere.  This is really equivalent to the: target.breakpoints-use-platform-avoid-list setting, which is also about "stop at places most normal human beings don't want to stop".  In the extant setting it's about whole libraries, the new setting would be about adding breakpoints on artificial line table entries or something like that.

Jim
 

> On Mar 4, 2024, at 3:45 PM, Felipe de Azevedo Piovezan ***@***.***> wrote:
> 
> 
> Also, are these 100% uninteresting line table entries? At present, there's no way to get lldb to not filter them out, so they either really need to be totally uninteresting, or we need a setting to turn them back on again.
> 
> They should be 100% uninteresting, but I agree that it would be nice to have a mechanism allowing these to be re-enabled.
> Do you think it is better to have a setting for this, or could we perhaps add the breakpoint but keep it disabled by default (is there precedent for this?) ?
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/llvm/llvm-project/pull/83908#issuecomment-1977664034>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ADUPVWYSKO52SMUY5VPMTF3YWUBRLAVCNFSM6AAAAABEF5NXPGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNZXGY3DIMBTGQ>.
> You are receiving this because your review was requested.
>

kastiglione commented on lldb/source/Breakpoint/BreakpointResolver.cpp:
```diff
@@ -199,13 +200,24 @@ bool operator<(const SourceLoc lhs, const SourceLoc rhs) {
     llvm::StringRef log_ident, uint32_t line, std::optional<uint16_t> column) {
   llvm::SmallVector<SymbolContext, 16> all_scs;
   for (uint32_t i = 0; i < sc_list.GetSize(); ++i)
     all_scs.push_back(sc_list[i]);
 
+  ApplyLanguageFilters(all_scs);
```
To repeat what @adrian-prantl said, what do you think about filtering inside this for loop, instead of filtering after the list has been constructed?

felipepiovezan replied:
Yup, I'll change that in the next iteration of this patch (it was uploaded before Adrian's comment)

felipepiovezan committed:
fixup! add target option

felipepiovezan commented:
I believe I've addressed all the feedback here.
Tried really hard to come up with a shorter name for that target option, but could not find something both short and descriptive. If you can think of anything better, please let me know

felipepiovezan committed:
fixup! rename property

felipepiovezan committed:
fixup! use range-based for for sc_list

felipepiovezan commented:
address review comments from Adrian.
I think I have a better name for the target option and language method now

felipepiovezan commented on lldb/include/lldb/Target/Language.h:
```diff
@@ -339,6 +339,15 @@ class Language : public PluginInterface {
+  /// create artificial functions without any meaningful user code associated
+  /// with them (e.g. code that gets expanded in late compilation stages, like
+  /// by CoroSplitter).
+  virtual bool IgnoreForLineBreakpoints(const SymbolContext &) const {
```
brevity: 
This is useful for -> Helpful for
any meaningful -> meaningful

felipepiovezan committed:
fixup! use language-based setting

jimingham reviewed (approved):
LGTM

felipepiovezan committed:
fixup! make language properties global

felipepiovezan committed:
fixup! brevity in comment language
